File: app/src/main/java/com/tetsoft/typego/SpannableEditText.java
package com.tetsoft.typego;

import android.content.Context;
import android.graphics.Color;
import android.text.Spannable;
import android.text.style.BackgroundColorSpan;
import android.text.style.ForegroundColorSpan;
import android.text.style.StyleSpan;
import android.util.AttributeSet;


public class SpannableEditText extends androidx.appcompat.widget.AppCompatEditText {
    private static final int GREEN_COLOR = Color.rgb(0, 128, 0);
    private static final int RED_COLOR = Color.rgb(192, 0, 0);
    private static final int EXCLUSIVE = Spannable.SPAN_EXCLUSIVE_EXCLUSIVE;

    public SpannableEditText(Context context) {
        super(context);
    }

    public SpannableEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public SpannableEditText(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public void paintForeground(int charIndex, ForegroundColorSpan foregroundColor) {
        if (this.getText() == null) return;
        if (charIndex + 1 > this.getText().length()) return;
        this.getText().setSpan(foregroundColor, charIndex, charIndex + 1, EXCLUSIVE);
    }

    public void paintForeground(int charIndex, int endIndex, ForegroundColorSpan foregroundColor) {
        if (this.getText() == null) return;
        if (charIndex + 1 > this.getText().length()) return;
        this.getText().setSpan(foregroundColor, charIndex, endIndex, EXCLUSIVE);
    }

    public void paintBackground(int startIndex, int endIndex, BackgroundColorSpan backgroundColor) {
        if (this.getText() == null) return;
        if (startIndex + 1 > this.getText().length() || endIndex + 1 > this.getText().length())
            return;
        this.getText().setSpan(backgroundColor, startIndex, endIndex, EXCLUSIVE);
    }

    public void clearForeground(int startIndex, int endIndex) {
        if (this.getText() == null) return;
        final Object[] removableSpans = this.getText().getSpans(startIndex, endIndex, Object.class);
        for (final Object span : removableSpans) {
            if (span instanceof ForegroundColorSpan) {
                this.getText().removeSpan(span);
            }
        }
    }

    public void clearBackground(int startIndex, int endIndex) {
        if (this.getText() == null) return;
        final Object[] removableSpans = this.getText().getSpans(startIndex, endIndex, Object.class);
        for (final Object span : removableSpans) {
            if (span instanceof BackgroundColorSpan || span instanceof StyleSpan || span instanceof ForegroundColorSpan) {
                this.getText().removeSpan(span);
            }
        }
    }

    public static ForegroundColorSpan getGreenForeground() {
        return new ForegroundColorSpan(GREEN_COLOR);
    }

    public static ForegroundColorSpan getRedForeground() {
        return new ForegroundColorSpan(RED_COLOR);
    }
}

File: app/src/main/java/com/tetsoft/typego/ui/custom/SpannableEditText.kt
package com.tetsoft.typego.ui.custom

import android.content.Context
import android.graphics.Color
import androidx.appcompat.widget.AppCompatEditText
import android.text.style.ForegroundColorSpan
import android.text.style.BackgroundColorSpan
import android.text.Spannable
import android.util.AttributeSet
import kotlin.math.min

class SpannableEditText : AppCompatEditText {
    constructor(context: Context?) : super(context!!)

    constructor(context: Context?, attrs: AttributeSet?) : super(
        context!!, attrs
    )

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(
        context!!, attrs, defStyle
    )

    var selectedWordStartPosition : Int = 0

    var selectedWordEndPosition : Int = 0

    var selectedWord : String = ""

    var autoScrollPosition = 0

    var autoScrollPredictPosition = 0

    fun selectNextWord() {
        val backgroundSpan = BackgroundColorSpan(Color.rgb(0, 80, 100))
        val foregroundSpan = ForegroundColorSpan(Color.WHITE)
        paintBackground(selectedWordStartPosition, selectedWordEndPosition, backgroundSpan)
        paintForeground(selectedWordStartPosition, selectedWordEndPosition, foregroundSpan)
        updateAutoScrollPosition()
    }

    private fun updateAutoScrollPosition() {
        autoScrollPosition = selectedWordEndPosition + autoScrollPredictPosition
        setSelection(min(autoScrollPosition, text!!.length))
    }

    fun setNextWordCursors() {
        selectedWordStartPosition = selectedWordEndPosition + 1
        selectedWordEndPosition = selectedWordStartPosition
        while (selectedWordEndPosition < length() && text?.get(selectedWordEndPosition) != ' ') {
            selectedWordEndPosition++
        }
        updateCurrentWord()
    }

    private fun updateCurrentWord() {
        if (selectedWordStartPosition > length() || selectedWordEndPosition > length()) return
        selectedWord = text.toString().substring(selectedWordStartPosition, selectedWordEndPosition)
    }

    fun initializeWordCursor() {
        while (selectedWordEndPosition < length() && text?.get(selectedWordEndPosition) != ' ') {
            selectedWordEndPosition++
        }
        updateCurrentWord()
    }

    fun deselectCurrentWord() {
        clearBackground(selectedWordStartPosition, selectedWordEndPosition)
    }

    fun selectCurrentWordAsIncorrect() {
        paintForeground(selectedWordStartPosition, selectedWordEndPosition, redForeground)
    }

    fun selectCurrentWordAsCorrect() {
        paintForeground(selectedWordStartPosition, selectedWordEndPosition, greenForeground)
    }

    fun reachedTheEnd() : Boolean {
        return selectedWordStartPosition >= text!!.length
    }

    fun paintForeground(charIndex: Int, foregroundColor: ForegroundColorSpan?) {
        if (this.text == null) return
        if (charIndex + 1 > this.text!!.length) return
        this.text!!.setSpan(foregroundColor, charIndex, charIndex + 1, EXCLUSIVE)
    }

    fun paintForeground(charIndex: Int, endIndex: Int, foregroundColor: ForegroundColorSpan?) {
        if (this.text == null) return
        if (charIndex + 1 > this.text!!.length || endIndex + 1 > this.text!!.length) return
        this.text!!.setSpan(foregroundColor, charIndex, endIndex, EXCLUSIVE)
    }

    fun paintBackground(startIndex: Int, endIndex: Int, backgroundColor: BackgroundColorSpan?) {
        if (this.text == null) return
        if (startIndex + 1 > this.text!!.length || endIndex + 1 > this.text!!.length) return
        this.text!!.setSpan(backgroundColor, startIndex, endIndex, EXCLUSIVE)
    }

    fun clearForeground(startIndex: Int, endIndex: Int) {
        if (this.text == null) return
        val removableSpans = this.text!!
            .getSpans(startIndex, endIndex, Any::class.java)
        for (span in removableSpans) {
            if (span is ForegroundColorSpan) {
                this.text!!.removeSpan(span)
            }
        }
    }

    fun clearBackground(startIndex: Int, endIndex: Int) {
        if (this.text == null) return
        val removableSpans = this.text!!
            .getSpans(startIndex, endIndex, Any::class.java)
        for (span in removableSpans) {
            if (span is BackgroundColorSpan) {
                this.text!!.removeSpan(span)
            }
        }
    }

    companion object {
        private val GREEN_COLOR = Color.rgb(0, 128, 0)
        private val RED_COLOR = Color.rgb(192, 0, 0)
        private const val EXCLUSIVE = Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
        @JvmStatic
        val greenForeground: ForegroundColorSpan
            get() = ForegroundColorSpan(GREEN_COLOR)
        @JvmStatic
        val redForeground: ForegroundColorSpan
            get() = ForegroundColorSpan(RED_COLOR)
    }
}